notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,372 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "841a8fc2-2efa-4327-b35c-e64a8fa380d3",
+   "metadata": {},
+   "source": [
+    "## Robotic updates to wikipedia by channel\n",
+    "If you want to determine which channels receive the most robotic changes you could use a regular aggregate query like this one:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79bac420-4ee7-4f77-abdf-be651e5752ef",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "  \"channel\",\n",
+    "  \"isRobot\",\n",
+    "  SUM(added) total_added,\n",
+    "  SUM(deleted) total_deleted\n",
+    "FROM \"example-pivot-unpivot\" \n",
+    "GROUP BY 1, 2\n",
+    "'''\n",
+    "display_client.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a8642f39-ad04-45b3-9120-b622a0f7c8a7",
+   "metadata": {},
+   "source": [
+    "## PIVOT example\n",
+    "The prior result is a bit long and makes it hard to figure out which channels have the most updates while still seeing the distinction between additions and deletions to the wikipedia pages.\n",
+    "One way to reduce the result while still keeping all the detail is to pivot the result such that rows become columns for each value of a particular column, the pivot column.\n",
+    "\n",
+    "PIVOT is a SQL operator that is used after the FROM clause.\n",
+    "\n",
+    "```\n",
+    "    FROM \"example-pivot-unpivot\" \n",
+    "    \n",
+    "    PIVOT ( SUM(added) as added, \n",
+    "            SUM(deleted) as deleted \n",
+    "            \n",
+    "            FOR \"isRobot\" IN ('true' as robot, 'false' as human)\n",
+    "          )\n",
+    "```\n",
+    "- It lists a set of aggregations to calculate: `SUM(added) as added, SUM(deleted) as deleted`\n",
+    "- The FOR clause indicated the pivot column: `\"isRobot\"`\n",
+    "- The IN clause lists the pivot values of the column and what to call them: `IN ('true' as robot, 'false' as human)`\n",
+    "\n",
+    "The resulting columns from this operation are a combination in the form \\<pivot_value_name\\>__\\<aggregation_name\\>. So in this case the result pivot columns will be:\n",
+    "```\n",
+    "\"robot_added\"\n",
+    "\"human_added\"\n",
+    "\"robot_deleted\"\n",
+    "\"human_deleted\"\n",
+    "```\n",
+    "which can now form part of the SELECT clause in the query.\n",
+    "\n",
+    "\n",
+    "The following query uses the results of the pivot to then aggregate totals for the pivot columns by channel. The aggregations being calculated could be NULL in the case where \"isRobot\"='true' (or 'false') does not exist for a given channel. The query calculates a ratio of human to total changes for each channel and the result is sorted on this ratio such that the channels with the highest proportion of robot updates are listed first. So in order to calculate this metric even in presence of NULLs the query uses COALESCE on all the pivoted metrics.\n",
```

Comment:
I find this paragraph hard to comprehend. Consider simplifying and using active voice.

Reply:
changed it